crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -863,27 +861,35 @@ where
             }
 
             if !new_segment_headers.is_empty() {
-                let maybe_block_number_to_finalize = {
-                    let mut segment_headers = segment_headers.lock();
-                    segment_headers.put(block_number + One::one(), new_segment_headers);
+                segment_headers
+                    .lock()
+                    .put(block_number + One::one(), new_segment_headers);
 
+                let maybe_block_number_to_finalize = segment_headers_store
+                    .max_segment_index()
                     // Skip last `FINALIZATION_DEPTH_IN_SEGMENTS` archived segments
-                    segment_headers
-                        .iter()
-                        .flat_map(|(_k, v)| v.iter().rev())
-                        .nth(FINALIZATION_DEPTH_IN_SEGMENTS.get())
-                        .map(|segment_header| segment_header.last_archived_block().number)
-                };
+                    .and_then(|max_segment_index| {
+                        max_segment_index.checked_sub(FINALIZATION_DEPTH_IN_SEGMENTS)
+                    })
+                    .and_then(|segment_index| {
+                        segment_headers_store.get_segment_header(segment_index)
+                    })
+                    .map(|segment_header| segment_header.last_archived_block().number)
+                    // Make sure not to finalize block number that does not yet exist (segment
+                    // headers store may contain future blocks during initial sync)
+                    .map(|block_number| best_archived_block_number.min(block_number.into()))
```

Comment:
It seems during the initial sync we may finalize the block as soon as it is K blocks deep. Is it possible that the block is then pruned and when the next block is imported, we will fail to query with `parent_block_hash` as it is pruned?

Reply:
Parent block of the best block will always exist, parent block of some block at the K depth can potentially disappear. Practically though it will still be present with recommended parameters because with `--blocks-pruning 256` it will prune blocks that are 256 below finalized, which we will handle separately at some point. Since these blocks are already deep in the history, we have no need in them and can finalize/prune quite quickly.

Is there something that would need to query those blocks during initial sync?

Reply:
>Is there something that would need to query those blocks during initial sync?

What I referred to is https://github.com/subspace/subspace/blob/ae1995e37c2a2680de6f8b9cd3390fc2e778dad2/crates/sc-consensus-subspace/src/archiver.rs#L827-L831

>Practically though it will still be present with recommended parameters because with --blocks-pruning 256 it will prune blocks that are 256 below finalized

Yeah, it is not a big issue, the user can restart the node to workaround the issue.